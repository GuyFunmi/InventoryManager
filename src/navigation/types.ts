import { NavigationProp, RouteProp, useNavigation } from '@react-navigation/native';
import { ItemType } from "../ui/screens";

export type RootStackParamList = {
    EditInventory: {item: ItemType};
    InventoryList: undefined;
    CreateInventory: undefined;
}

// Type for useNavigation hook
export type NavigationProps = NavigationProp<RootStackParamList>;

// Type for navigation prop in components
export type ScreenNavigationProp<T extends keyof RootStackParamList> = NavigationProp<
  RootStackParamList,
  T
>;

// Type for route prop in components
export type ScreenRouteProp<T extends keyof RootStackParamList> = RouteProp<
  RootStackParamList,
  T
>;

// 4. Create type-safe props interface for screens
export interface ScreenProps<T extends keyof RootStackParamList> {
  navigation: ScreenNavigationProp<T>;
  route: ScreenRouteProp<T>;
}

// 6. Type-safe navigation hook usage
export const useAppNavigation = () => useNavigation<NavigationProps>();