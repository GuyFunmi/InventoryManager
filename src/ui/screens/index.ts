export * from "./CreateInventory";
export * from "./EditInventory";
export * from "./InventoryList";