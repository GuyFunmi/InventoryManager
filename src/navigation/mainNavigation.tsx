import React from 'react';
import { CardStyleInterpolators, createStackNavigator } from '@react-navigation/stack';
import { InventoryList, CreateInventory, EditInventory } from '../ui/screens';
import { RootStackParamList } from './types';

const Stack = createStackNavigator<RootStackParamList>();

const MainNavigation = () => {
  return (
      <Stack.Navigator screenOptions={details => {
        const {route} = details;
        return {
          cardStyleInterpolator: CardStyleInterpolators.forHorizontalIOS,
          headerShown: false,
          animationEnabled: false,
        };
      }}>
        <Stack.Screen name="InventoryList" component={InventoryList} />
        <Stack.Screen name="CreateInventory" component={CreateInventory} />
        <Stack.Screen name="EditInventory" component={EditInventory} />
      </Stack.Navigator>
  );
};

export default MainNavigation;
