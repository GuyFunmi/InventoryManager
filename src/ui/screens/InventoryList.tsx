import { useFocusEffect } from '@react-navigation/native';
import React, { memo, useState, useCallback } from 'react';
import { View, Text, FlatList, StyleSheet, TouchableOpacity } from 'react-native';
import { ScreenProps } from '../../navigation/types';
import { getItem } from '../../utils/helpers';
import { Layout } from '../components/Layout';

export type ItemType = {
  name: string;
  price: string;
  stock: string;
  description: string;
};

export const InventoryList = memo(({navigation}: ScreenProps<"InventoryList">) => {
  const [inventory, setInventory] = useState<ItemType[]>([]);

  useFocusEffect(
    useCallback(() => {
      const fetchData = async () => {
        const data = await getItem('inventory');
        if (data) setInventory(data);
      };

      fetchData();
    }, [navigation])
  );

  return (
    <Layout title={'Inventory List'}>
      <View style={styles.container}>
        <FlatList
          data={inventory}
          keyExtractor={(item) => item.name}
          renderItem={({ item }) => (
            <View style={styles.itemContainer}>
              <Text style={styles.itemName}>Name: {item.name}</Text>
              <Text style={styles.itemDetail}>Price: {item.price}</Text>
              <Text style={styles.itemDetail}>Total Stock: {item.stock}</Text>
              <TouchableOpacity
                style={styles.editButton}
                onPress={() => navigation.navigate('EditInventory', { item })}
              >
                <Text style={styles.editButtonText}>Edit</Text>
              </TouchableOpacity>
            </View>
          )}
        />
        <TouchableOpacity
          style={styles.addButton}
          onPress={() => navigation.navigate('CreateInventory')}
        >
          <Text style={styles.addButtonText}>Add Item</Text>
        </TouchableOpacity>
      </View>
    </Layout>
  );
});

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
    backgroundColor: '#f9f9f9',
  },
  itemContainer: {
    backgroundColor: '#fff',
    padding: 16,
    borderRadius: 8,
    marginBottom: 12,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 2,
  },
  itemName: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 4,
  },
  itemDetail: {
    fontSize: 16,
    color: '#555',
    marginBottom: 8,
  },
  editButton: {
    backgroundColor: '#4CAF50',
    paddingVertical: 8,
    paddingHorizontal: 16,
    borderRadius: 8,
  },
  editButtonText: {
    color: '#fff',
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  addButton: {
    backgroundColor: '#2196F3',
    paddingVertical: 12,
    paddingHorizontal: 24,
    borderRadius: 8,
    marginTop: 16,
    alignItems: 'center',
  },
  addButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
});
