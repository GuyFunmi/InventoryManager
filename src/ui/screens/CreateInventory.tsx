import React, { useState, memo } from 'react';
import { View, Text, TextInput, Button, Alert, StyleSheet, TouchableOpacity } from 'react-native';
import { ScreenProps, useAppNavigation } from '../../navigation/types';
import { storeItem, getItem } from '../../utils/helpers';
import { ItemType } from './InventoryList';
import {Layout} from '../components/Layout';

export const CreateInventory: React.FC<ScreenProps<"CreateInventory">> = memo(({navigation}) => {
  const [name, setName] = useState('');
  const [price, setPrice] = useState('');
  const [stock, setStock] = useState('');
  const [description, setDescription] = useState('');

  const validateForm = async () => {
    if (!name || !price || !stock || !description) {
      Alert.alert('All fields are required.');
      return false;
    }

    const existingInventory = await getItem('inventory');
    if (existingInventory?.some((item: ItemType) => item.name === name)) {
      Alert.alert('Name must be unique.');
      return false;
    }

    if (isNaN(Number(price)) || isNaN(Number(stock))) {
      Alert.alert('Price and Stock must be numbers.');
      return false;
    }

    if (description.split(' ')?.length < 3) {
      Alert.alert('Description must have at least 3 words.');
      return false;
    }

    return true;
  };

  const handleSubmit = async () => {
    const isValid = await validateForm();
    if (!isValid) return;

    const newItem = { name, price, stock, description };
    const inventory = (await getItem('inventory')) || [];
    inventory.push(newItem);
    await storeItem('inventory', inventory);
    navigation.navigate('InventoryList');
  };

  return (
    <Layout title={'Create Inventory'}>
      <View style={styles.container}>
        <TextInput
          style={styles.input}
          placeholder="Name"
          value={name}
          onChangeText={setName}
        />
        <TextInput
          style={styles.input}
          placeholder="Price"
          value={price}
          onChangeText={setPrice}
          keyboardType="numeric"
        />
        <TextInput
          style={styles.input}
          placeholder="Total Stock"
          value={stock}
          onChangeText={setStock}
          keyboardType="numeric"
        />
        <TextInput
          style={styles.input}
          placeholder="Description"
          value={description}
          onChangeText={setDescription}
          multiline
        />
        <TouchableOpacity
          style={styles.button}
          onPress={handleSubmit}
        >
          <Text style={styles.buttonText}>Save</Text>
        </TouchableOpacity>
      </View>
    </Layout>
  );
});

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#2196F3',
    paddingVertical: 12,
    paddingHorizontal: 24,
    borderRadius: 8,
    marginTop: 16,
    alignItems: 'center',
  },
  buttonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
  container: {
    flex: 1,
    padding: 20,
    backgroundColor: '#f9f9f9',
  },
  input: {
    height: 50,
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 16,
    marginBottom: 12,
    backgroundColor: '#fff',
  },
});
