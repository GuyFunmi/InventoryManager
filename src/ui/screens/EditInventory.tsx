import React, { useState, memo } from 'react';
import { View, Text, TextInput, Alert, StyleSheet, TouchableOpacity } from 'react-native';
import { getItem, updateItem, deleteItem } from '../../utils/helpers';
import { ScreenProps } from '../../navigation/types';
import { ItemType } from './InventoryList';
import {Layout} from '../components/Layout';

export const EditInventory: React.FC<ScreenProps<"EditInventory">> = memo(({navigation, route}) => {
  const item = (route as { params: { item: ItemType } })?.params?.item;

  const [name, setName] = useState(item.name);
  const [price, setPrice] = useState(item.price);
  const [stock, setStock] = useState(item.stock);
  const [description, setDescription] = useState(item.description);
  const [originalName, setOriginalName] = useState(item.name);

  const validateForm = async () => {
    if (!name || !price || !stock || !description) {
      Alert.alert('All fields are required.');
      return false;
    }

    const existingInventory: ItemType[] = await getItem('inventory');
    if (name !== originalName && existingInventory?.some((inv) => inv.name === name)) {
      Alert.alert('Name must be unique.');
      return false;
    }

    if (isNaN(Number(price)) || isNaN(Number(stock))) {
      Alert.alert('Price and Stock must be numbers.');
      return false;
    }

    if (description.split(' ').length < 3) {
      Alert.alert('Description must have at least 3 words.');
      return false;
    }

    return true;
  };

  const handleSave = async () => {
    const isValid = await validateForm();
    if (!isValid) return;

    const updatedItem = { name, price, stock, description };
    const inventory: ItemType[] = await getItem('inventory');
    const updatedInventory = inventory.map((inv) =>
      inv.name === originalName ? updatedItem : inv
    );
    await updateItem('inventory', updatedInventory);

    navigation.navigate('InventoryList');
  };

  const handleDelete = async () => {
    Alert.alert(
      'Confirm Delete',
      'Are you sure you want to delete this item?',
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'OK',
          onPress: async () => {
            const inventory: ItemType[] = await getItem('inventory');
            const updatedInventory = inventory.filter((inv) => inv.name !== originalName);
            await deleteItem('inventory');
            await updateItem('inventory', updatedInventory);

            navigation.navigate('InventoryList');
          },
        },
      ],
      { cancelable: true }
    );
  };

  return (
    <Layout showBackButton title={'Edit Inventory'}>
      <View style={styles.container}>
        <TextInput
          style={styles.input}
          placeholder="Name"
          value={name}
          onChangeText={setName}
        />
        <TextInput
          style={styles.input}
          placeholder="Price"
          value={price.toString()}
          onChangeText={setPrice}
          keyboardType="numeric"
        />
        <TextInput
          style={styles.input}
          placeholder="Total Stock"
          value={stock.toString()}
          onChangeText={setStock}
          keyboardType="numeric"
        />
        <TextInput
          style={styles.input}
          placeholder="Description"
          value={description}
          onChangeText={setDescription}
          multiline
        />

        <TouchableOpacity
          style={styles.saveButton}
          onPress={handleSave}
        >
          <Text style={styles.saveButtonText}>Save</Text>
        </TouchableOpacity>

        <TouchableOpacity
          style={styles.deleteButton}
          onPress={handleDelete}
        >
          <Text style={styles.deleteButtonText}>Delete</Text>
        </TouchableOpacity>
      </View>
    </Layout>
  );
});

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    backgroundColor: '#f5f5f5',
  },
  deleteButton: {
    paddingVertical: 12,
    borderWidth: 0.2,
    borderColor: '#3c3c3c',
    paddingHorizontal: 24,
    borderRadius: 8,
    marginTop: 16,
    alignItems: 'center',
  },
  deleteButtonText: {
    color: '#ff4d4d',
    fontSize: 18,
    fontWeight: 'bold',
  },
  input: {
    height: 50,
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 16,
    marginBottom: 16,
    backgroundColor: '#fff',
  },
  saveButton: {
    backgroundColor: '#2196F3',
    paddingVertical: 12,
    paddingHorizontal: 24,
    borderRadius: 8,
    marginTop: 16,
    alignItems: 'center',
  },
  saveButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
});
