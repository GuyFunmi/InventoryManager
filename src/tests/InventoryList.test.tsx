import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { InventoryList } from '../ui/screens/InventoryList';
import React from 'react';
import { ScreenRouteProp } from '../navigation/types';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
  getAllKeys: jest.fn(),
}));

// Mock navigation
const mockNavigation: any = {
  navigate: jest.fn(),
  dispatch: jest.fn(),
  reset: jest.fn(),
  goBack: jest.fn(),
  isFocused: jest.fn(),
  canGoBack: jest.fn(),
  setParams: jest.fn(),
  setOptions: jest.fn(),
  getId: jest.fn(),
  addListener: jest.fn(),
  removeListener: jest.fn(),
  pop: jest.fn(),
  popToTop: jest.fn(),
};

// Mock route
const mockRoute: ScreenRouteProp<'InventoryList'> = {
  key: 'InventoryList-123',
  name: 'InventoryList',
  params: undefined,
};

// Mock Layout component
jest.mock('../ui/components/Layout', () => ({
  Layout: ({ children }: { children: React.ReactNode }) => <>{children}</>,
}));

const mockInventoryData = [
  { name: 'Glock', price: '50', stock: '500', description: 'Test Item 1' },
  { name: 'Baby', price: '500', stock: '200', description: 'Test Item 2' },
];

// Mock useFocusEffect to execute callback immediately
jest.mock('@react-navigation/native', () => {
  const actual = jest.requireActual('@react-navigation/native');
  return {
    ...actual,
    useFocusEffect: (callback: any) => {
      callback();
    },
    useNavigation: () => mockNavigation,
  };
});

describe('InventoryList Screen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    // Setup AsyncStorage mock to return our test data
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify(mockInventoryData));
  });

  it('should render inventory items and match snapshot', async () => {
    const { findByText, toJSON } = render(
      <InventoryList navigation={mockNavigation} route={mockRoute} />
    );

    // Wait for an item to appear
    await findByText('Name: Glock');
    expect(toJSON()).toMatchSnapshot();
  });

  it('should call AsyncStorage.getItem on mount', async () => {
    render(<InventoryList navigation={mockNavigation} route={mockRoute} />);
    
    // Verify AsyncStorage.getItem was called
    expect(AsyncStorage.getItem).toHaveBeenCalledWith('inventory');
    
    // Wait for the async operation to complete
    await waitFor(() => {
      expect(AsyncStorage.getItem).toHaveBeenCalledTimes(1);
    });
  });

  it('should navigate to EditInventory when Edit button is pressed', async () => {
    const { findAllByText } = render(
      <InventoryList navigation={mockNavigation} route={mockRoute} />
    );
    // Wait for Edit buttons to appear
    const editButtons = await findAllByText('Edit');
    fireEvent.press(editButtons[0]);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('EditInventory', {
      item: mockInventoryData[0]
    });
  });

  it('should display all item details correctly', async () => {
    const { findByText } = render(
      <InventoryList navigation={mockNavigation} route={mockRoute} />
    );

    // Verify all the data is displayed
    await findByText('Name: Glock');
    await findByText('Price: 50');
    await findByText('Total Stock: 500');
    await findByText('Name: Baby');
    await findByText('Price: 500');
    await findByText('Total Stock: 200');
  });

  it('should handle empty inventory data', async () => {
    // Mock empty data for this test
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));

    const { queryByText, findByText } = render(
      <InventoryList navigation={mockNavigation} route={mockRoute} />
    );

    // Wait for component to render
    await findByText('Add Item');

    // Verify no items are rendered
    expect(queryByText(/Name: /)).toBeNull();
  });

  it('should handle null inventory data', async () => {
    // Mock null data for this test
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(null);

    const { queryByText, findByText } = render(
      <InventoryList navigation={mockNavigation} route={mockRoute} />
    );

    // Wait for component to render
    await findByText('Add Item');

    // Verify no items are rendered
    expect(queryByText(/Name: /)).toBeNull();
  });
});