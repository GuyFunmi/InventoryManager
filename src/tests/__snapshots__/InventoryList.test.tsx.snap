// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditInventory Screen should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#f5f5f5",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Name"
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 50,
        "marginBottom": 16,
        "paddingHorizontal": 16,
      }
    }
    value="Test Item"
  />
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Price"
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 50,
        "marginBottom": 16,
        "paddingHorizontal": 16,
      }
    }
    value="10.99"
  />
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="Total Stock"
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 50,
        "marginBottom": 16,
        "paddingHorizontal": 16,
      }
    }
    value="100"
  />
  <TextInput
    multiline={true}
    onChangeText={[Function]}
    placeholder="Description"
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 50,
        "marginBottom": 16,
        "paddingHorizontal": 16,
      }
    }
    value="Test description"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2196F3",
        "borderRadius": 8,
        "marginTop": 16,
        "opacity": 1,
        "paddingHorizontal": 24,
        "paddingVertical": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Save
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderColor": "#3c3c3c",
        "borderRadius": 8,
        "borderWidth": 0.2,
        "marginTop": 16,
        "opacity": 1,
        "paddingHorizontal": 24,
        "paddingVertical": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ff4d4d",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Delete
    </Text>
  </View>
</View>
`;

exports[`InventoryList Screen should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9f9f9",
      "flex": 1,
      "padding": 16,
    }
  }
>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2196F3",
        "borderRadius": 8,
        "marginTop": 16,
        "opacity": 1,
        "paddingHorizontal": 24,
        "paddingVertical": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Add Item
    </Text>
  </View>
</View>
`;
